### PR TITLE
Updating language_reference

### PR DESCRIPTION
**Pull request by ewoudje**
While making a intelij lang plugin (very WIP), I noticed that the language reference is missing some things.

Changes:
 * Making some lines less ambiguous
 * Adding missing parts that i did test with the latest windows release

### CODE REVIEW

File: docs/source/language_reference.md [hunk 1]
```diff
@@ -51,8 +51,8 @@ Lexical definition
 Grammar
 -------
 
-Below, `...` indicates a loop with exit point at that scope level, and `||` is
-like `|` except indicates a precedence level difference.
+Below, `...` indicates a loop with exit point at that scope level (ex. `(ident ... ,)` -> `(ident (, ident)*)`, * meaning optionaly repeating), 
+and `||` islike `|` except indicates a precedence level difference. `[rule]` Means optional.
```
Comment (aardappel) [inline docs/source/language_reference.md:55]: space between `is` and `like` ?

File: docs/source/language_reference.md [hunk 2]
```diff
@@ -64,54 +64,54 @@ topexp = `namespace` ident
       \| expstat \| attrdef
 
 class = ( `class` \| `struct` ) ident
-        ( `=` ident specializers
-       \| [ generics ] `:` [ ident [ specializers ] ]
+        ( ( `=` ident specializers )
```
Comment (aardappel) [inline docs/source/language_reference.md:67]: these are not required

File: docs/source/language_reference.md [hunk 3]
```diff
@@ -64,54 +64,54 @@ topexp = `namespace` ident
...
+       \| ( [ generics ] `:` [ ident [ specializers ] ] )
           indlist( ident [ `:` type ] [ `=` exp ] \| functiondef ) )
 
 specializers = `<` list( type ) `>`
 
 generics = `<` list( ident ) `>`
 
-vardef = ( `var` \| `let` ) list( ident ) `=` opexp
+vardef = ( `var` \| `let` ) list( ident ) [ `:` type ] `=` opexp
```
Comment (aardappel) [inline docs/source/language_reference.md:75]: type goes inside the list

File: docs/source/language_reference.md [hunk 4]
```diff
@@ -64,54 +64,54 @@ topexp = `namespace` ident
...
 enumdef = ( `enum` | `enum_flags` ) ident `:` indlist( ident [ `=` integer\_constant ] )
 
-functiondef = `def` ident generics functionargsbody
+functiondef = `def` ident [ generics ] functionargsbody
 
-functionargsbody = `(` args `) :` body
+functionargsbody = `(` args `)` `:` body
 
-block = [ args ] `:` body \| functionargsbody
+block = ( [ args ] `:` body ) \| functionargsbody
```
Comment (aardappel) [inline docs/source/language_reference.md:83]: not needed

File: docs/source/language_reference.md [hunk 5]
```diff
@@ -64,54 +64,54 @@ topexp = `namespace` ident
...
 
-args = [ list( ident [ ( `:` \| `::` ) type ] ) ]
+args = [ list( ident [ ( `:` \| `::` ) type ] [ `=` exp ] ) ]
 
-body = ( expstat \| indent stats dedent )
+body = ( expstat \| ( indent stats dedent ) )
```
Comment (aardappel) [inline docs/source/language_reference.md:87]: not needed

File: docs/source/language_reference.md [hunk 6]
```diff
@@ -64,54 +64,54 @@ topexp = `namespace` ident
...
 
-type = `int` \| `float` \| `string` \| `[` type `]` \| `resource` `<` ident `>` \| `void`
-    \| ident
+type = `int` \| `float` \| `string` \| ( `[` type `]` ) \| ( `resource` `<` ident `>` ) \| `void`
```
Comment (aardappel) [inline docs/source/language_reference.md:89]: please remove all unnecessary ()

File: docs/source/language_reference.md [hunk 7]
```diff
@@ -64,54 +64,54 @@ topexp = `namespace` ident
...
+    \| (ident specializers) \| ident
```
Comment (aardappel) [inline docs/source/language_reference.md:90]: instead put `[ specializers ]` behind the existing `ident`

File: docs/source/language_reference.md [hunk 8]
```diff
@@ -64,54 +64,54 @@ topexp = `namespace` ident
...
 
-call = specializers `(` [ list( exp ) ] `)` [ block [ `fn` block … ] ]
+call = [ specializers ] `(` [ list( exp ) ] `)` [ block [ `fn` block … ] ]
 
-expstat = ( exp … `;` ) \| `return` ( [ list( opexp ) ] ) [ `from` ( `program`
-\| ident ) ]
+expstat = ( exp … `;` ) \| ( `return` ( [ list( opexp ) ] ) [ `from` ( `program`
```
Comment (aardappel) [inline docs/source/language_reference.md:94]: remove

File: docs/source/language_reference.md [hunk 9]
```diff
@@ -64,54 +64,54 @@ topexp = `namespace` ident
...
 exp = opexp [ ( `=` \| `+=` \| `-=` \| `*=` \| `/=` \| `%=` ) exp ]
 
-opexp = unary ( `*` \| `/` \| `%` \|\| `+` \| `-` \|\| `<` \| `>` \| `>=` \|
+opexp = unary [ ( `*` \| `/` \| `%` \|\| `+` \| `-` \|\| `<` \| `>` \| `>=` \|
 `<=` \|\| `==` \| `!=` \|\| `&` \| `|` \| `and` \| `or` \| \^ \|
-`<<` \| `>>`) unary
+`<<` \| `>>`) unary ]
 
-unary = ( `-` \| `++` \| `--` \| \~ \| `not` ) unary \| deref
+unary = ( ( `-` \| `++` \| `--` \| \~ \| `not` ) unary ) \| deref
```
Comment (aardappel) [inline docs/source/language_reference.md:103]: remove

File: docs/source/language_reference.md [hunk 10]
```diff
@@ -64,54 +64,54 @@ topexp = `namespace` ident
...
 
-deref = factor [ `[` exp `]` \| `.` ident [ call ] \| `->` ident
-\| `++` \| `--` \| `is` type ]
+deref = factor [ ( `[` exp `]` ) \| ( `.` ident [ call ] ) \| ( `->` ident )
```
Comment (aardappel) [inline docs/source/language_reference.md:105]: remove

File: docs/source/language_reference.md [hunk 11]
```diff
@@ -64,54 +64,54 @@ topexp = `namespace` ident
...
+\| `++` \| `--` \| ( `is` typ e) ]
```
Comment (aardappel) [inline docs/source/language_reference.md:106]: `type`

File: docs/source/language_reference.md [hunk 12]
```diff
@@ -64,54 +64,54 @@ topexp = `namespace` ident
...
 
 factor = constant \| `(` exp `)` \| constructor \| `fn` functionargsbody \|
-ident [ call ]
+( ident [ call ] )
```
Comment (aardappel) [inline docs/source/language_reference.md:109]: remove

File: docs/source/language_reference.md [hunk 13]
```diff
@@ -64,54 +64,54 @@ topexp = `namespace` ident
...
 
-constructor = `[` [ list( exp ) ] `]` [ `::` type ] \| ident `{` [ list(
+constructor = `[` [ list( exp ) ] `]` [ `::` type ] \| ident `{` [ list( ( ident `:` exp ) \| (
```
Comment (aardappel) [inline docs/source/language_reference.md:111]: remove ()

File: docs/source/language_reference.md [hunk 14]
```diff
@@ -64,54 +64,54 @@ topexp = `namespace` ident
...
 exp ) ] `}`
 
-constant = numeric\_constant \| string\_constant \| character\_constant \| `nil` [ `::` type ]
+constant = numeric\_constant \| string\_constant \| character\_constant \| ( `nil` [ `::` type ] )
```
Comment (aardappel) [inline docs/source/language_reference.md:114]: remove

**Comment by ewoudje** [comment]
With latest commit, i shouldve fixed all the comments.

**Comment by aardappel** [comment]
I still see some extra (), can you double check?

**Comment by ewoudje** [comment]
> I still see some extra (), can you double check?

Yep, seems like I forgot 1 and forgot also to remove a closing one.
Pretty sure I'm not missing any of them.

**Comment by aardappel** [comment]
Thanks!

**Pull request merged**